src/df.jl

```diff
@@ -45,18 +60,23 @@ end
 
 stringify(x) = filter(t -> t != ':', string(x))
 
-select_column(df, s) = haskey(df, s) ? convert_column(df[s]) : s
-
-convert_column(col) = col
+compute_name(df, i) = column_names(IterableTables.getiterator(df))[i]
 
-function convert_column(col::AbstractDataArray{T}) where T
-    try
-        convert(Array, col)
-    catch
-        error("Missing data of type $T is not supported")
+function compute_all(df, syms...)
+    iter = IterableTables.getiterator(df)
+    type_info = Dict(zip(column_names(iter), column_types(iter)))
+    cols = Tuple(isa(s, Integer) ? Array{column_types(iter)[s]}(0) :
+        s in column_names(iter) ? Array{type_info[s]}(0) : s for s in syms)
```

Comment:
Ah, so here you actually only materialize some columns, right? I think that means you should not use TableTraitsUtils.jl, because that will always materialize all columns.

One thing you could add later is to check whether ``Base.iteratorsize(typeof(iter))==Base.HasLength()``, in which case you can call ``length(iter)`` to get the number of rows you will eventually get and then you can either pre-allocate the array or maybe just use ``sizehint!`` on the arrays that you allocate here. In general that should speed up the ``for`` loop later where you push into the array.

Reply:
Yes, essentially the idea is for the df macro to be a with-style macro that only extracts the columns that are actually needed (and only tries to do this on symbols that are names in the table).

Reply:
I should probably check all the techniques you use in TableTraitsUtils.jl to speed up this part (it seems to me that mainly I'm missing unlooping the `for` loop over column names with generated functions and preallocating wherever possible). Still, this feels like unnecessary code duplication. By looking at your code, it seems like it'd be very easy to generalize it to only return some of the columns, with an appropriate keyword argument (if you want I can try and make a PR there). In this case, I don't think this would go against the spirit of keeping IterableTables as tiny as possible, as developers of other packages wouldn't need to do anything extra to support this new feature, whereas the StatPlots use case (select a few columns from an IterableTable) seems common enough and I'm not sure the code to do that efficiently belongs to a statistical plotting package. What do you think?

Reply:
Yes, I thought a bit more about this and actually came to the same conclusion :) If we just keep that as an option in TableTraitsUtils.jl it is all cool as it won't complicate the core interface in TableTraits.jl at all.

Reply:
@davidanthoff This is the fist time I'm taking a deeper look at IterableTables (has been on my to-do list for a while though ;-) ). Maybe I didn't look in the right place in the docs - what is the recommended way to iterate over a subset of columns without loading all of them?

Reply:
If there isn't anything like that at the moment - @piever,  @davidanthoff, if I understand correctly, you're suggesting something like `getiterator(ds, cols = ...)`? That would be very handy in many other use cases (data analysis in general), indeed.

Reply:
I don't want to change the iterable tables interface, i.e. you would still always get an iterator that returns a named tuple that has all the columns in it. But I want to now provide a helper function in TableTraitsUtils.jl that materializes only a subset of the columns into arrays.

There are two reasons: one is that it keeps the iterable tables interface simple. Adding the ability to tell the source which columns it should return adds additional complexity requirements on the source, and it is a bit difficult to see where to stop. Plus, I think I have a vague idea how one can have a similar story based on something in Query.jl. Not fleshed out yet, but stay tuned ;)

The other reason is that I think the compiler might actually be able to optimize all the tuple access to columns that aren't used away. I'm not a 100% sure about that, but when I talked with Keno about that in July I got the impression that he might think that feasible. In general, if one is careful, a lot of things in the iterable tables interface get inlined and one typically ends up with just one tight loop with no function calls at all. If, inside that loop, one accesses a column, but then that column is never used inside the loop, it should in theory be feasible for the compiler to detect that and just remove that access. If that is true and works, there is actually no need to add an option to the interface itself to select columns. I think for now it is probably best to at some point check out that option carefully, and if that can't be pulled of we can revisit an option in the core iterable tables interface.

Reply:
> The other reason is that I think the compiler might actually be able to optimize all the tuple access to columns that aren't used away.

I had hoped that might be the case, but at least in a very simple toy example I tried it wasn't the case. Inlining is, of course, sometimes hard to predict/control.

But I was also thinking about out-of-core data sources (e.g. databases, column-store file-formats, etc.), where one may need up-front knowledge about which columns to load, as loading columns may be expensive. I think in Query.jl, that's currently hard-coded for SQLite, but a generic solution would be very cool.

I do understand the need to keep things simple, of course.

> typically ends up with just one tight loop with no function calls at all

Might that even SIMD-vectorize? Would depend on not having any branch-expressions, of course (so no on-demand loading, etc.).

Reply:
Sorry for getting out-of-topic a bit, here.

Reply:
> I had hoped that might be the case, but at least in a very simple toy example I tried it wasn't. Inlining is, of course, sometimes hard to predict/control.

I tried that way, way back and a lot of even quite involved Query.jl queries ended up being inlined  into one tight loop. I'm not sure whether that is enough for the compiler to ignore access to columns that are not needed, though...

> I think in Query.jl, that's currently hard-coded for SQLite, but a generic solution would be very cool.

Yes, totally agree. My current thinking is that it would be great if a source could implement a partial Query.jl backend. For example, say you have this code ``load("file.csv") |> @select({_.a, _.b}) |> DataFrame`` (*), it would first go to a CSVFiles.jl Query.jl backend that realizes that this query implies that only column ``a`` and ``b`` are needed, and then skips all other columns at CSV parse time already. The same logic could apply to lots of other sources. But for this to work it would be really key that CSVFiles.jl doesn't have to implement a full Query.jl backend, but some lighter version of it. I've been mulling this whole question for a while, and it is slowly clearing up in my mind, but will probably still take a while.

> Might that even SIMD-vectorize? Would depend on not having any branch-expressions, of course (so no on-demand loading, etc.).

Yeah, I think that could probably be made to work.

Reply:
Sounds exciting - I look forward to see what you'll come up with. I know how it is, sometimes concepts just need some time to mature in ones mind ...